### DISCUSSION TITLE
USB fan speed?

### DISCUSSION THREAD

**Discussion by mntbighker** [nicolargo/glances#2865] [Ideas]
I have a Pi5 with an ArgonOne case that uses a USB bus connected fan. So unlike the factory PWM fan, you won't see the fan speed in glances. Has anyone looked at adding this to glances? Here is a code snippet from the Argon script in case it's of use.

```
def argonregister_getfanspeed(busobj, regsupport=None):
        if busobj is None:
                return 0

        usereg=False
        if regsupport is None:
                usereg=argonregister_checksupport(busobj)
        else:
                usereg=regsupport
        if usereg == True:
                return argonregister_getbyte(busobj, ADDR_ARGONONEREG_DUTYCYCLE)
        else:
                return 0

```

Given how use case specific this is, it's likely not practical to add?

**Comment by nicolargo** [comment]
Complete Python script can be found here: https://github.com/kounch/argonone/blob/master/argononed.py but there is no API to get the Fan speed: https://github.com/kounch/argonone/blob/master/argononed.py#L88-L107

Perhaps the best solution to integrate it in Glances is by using AMP: https://glances.readthedocs.io/en/latest/aoa/amps.html

By adding something like that in the glances.conf file:

```
[amp_argonone_fan]
enable=false
refresh=30
one_line=false
command=python3 /folder/to/argononed.py
```

**Comment by mntbighker** [comment]
```
[amp_argonone_fan]
enable=true
refresh=30
one_line=false
command=python3 /etc/argon/argonstatus.py "fan speed"
```
This almost works, but the returned format isn't compatible it appears. Also, this gives you the duty cycle of the fan, and not the actual speed. In zsh this returns "Fan Speed x"

python3 /etc/argon/argonstatus.py "fan speed" | grep "Fan Speed"

So a one line bash wrapper works (pipe symbol verboten in AMP). This is pretty "hacky", but it gets the job done in case anyone else sees this. 

Thanks for the suggestion @nicolargo 

```
elif cmd == "fan speed":
         # Fan Speed
         try:
             newspeed = argonregister_getfanspeed(argonregister_initializebusobj())
             if newspeed <= 0:
                 fanconfig = load_fancpuconfig()
                 fanhddconfig = load_fanhddconfig()

                 # Speed based on CPU Temp
                 val = argonsysinfo_getcputemp()
                 newspeed = get_fanspeed(val, fanconfig)

                 val = argonsysinfo_getmaxhddtemp()
                 tmpspeed = get_fanspeed(val, fanhddconfig)
                 if tmpspeed > newspeed:
                     newspeed = tmpspeed
             print(baseleftoffset+"Fan Speed",str(newspeed))

```
